tfx/examples/imdb/imdb_utils_native_keras.py

```diff
@@ -0,0 +1,204 @@
+# Lint as: python2, python3
+# Copyright 2019 Google LLC. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Python source file include IMDB pipeline functions and necessary utils.
+
+The utilities in this file are used to build a model with native Keras.
+This module file will be used in Transform and generic Trainer."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from typing import List, Text
+
+import tensorflow as tf
+from tensorflow import keras
+import tensorflow_transform as tft
+
+from tfx.components.trainer.executor import TrainerFnArgs
+
+# There are 50,000 entries in the imdb dataset. ExampleGen splits the dataset
+# with a 2:1 train-eval ratio. Batch_size is an empirically sound
+# configuration.
+_TRAIN_BATCH_SIZE = 64
+_EVAL_BATCH_SIZE = 64
+_LABEL_KEY = "sentiment"
+_MAX_FEATURES = 8000
+_MAX_LEN = 100
+_LEARNING_RATE = 1e-4
+_TRAIN_EPOCHS = 10
+_DROPOUT_RATE = 0.2
+_HIDDEN_UNITS = 64
+
+def _gzip_reader_fn(filenames):
+  """Small utility returning a record reader that can read gzip'ed files."""
+  return tf.data.TFRecordDataset(filenames, compression_type='GZIP')
+
+def _tokenize_review(review):
+  """Tokenize the reivews by spliting the reviews, then constructing a
+  vocabulary. Map the words to their frequency index in the vocabulary."""
+  review_sparse = tf.strings.split(tf.reshape(review, [-1])).to_sparse()
+  # tft.apply_vocaublary doesn't reserve 0 for oov words. In order to comply
+  # with convention and use mask_zero in keras.embedding layer, manually set
+  # default value to -1 and add 1 to every index.
+  review_indices = tft.compute_and_apply_vocabulary(
+      review_sparse,
+      default_value=-1,
+      top_k=_MAX_FEATURES)
+  dense = tf.sparse.reset_shape(review_indices, None)
+  dense = tf.sparse.to_dense(review_indices, default_value=-1)
+  # TFX transform expect the transform result to be FixedLenFeature.
+  padding_config = [[0, 0], [0, _MAX_LEN]]
```

Comment:
Would the following work: padding_config = [[0, 0], [0, _MAX_LEN-tf.shape(dense)[1]]]

Then you would not need to slice after.

Reply:
No it won't work. Since tf.shape(dense)[1] is none.